### PR TITLE
feat: support time-based forking

### PR DESCRIPTION
**Pull request by onbjerg**
A revisit of #985 that adds support for time-based forks with network integration.

Makes hardfork activation checks more explicit, but also simplifies some parts of the chain spec (e.g. the duplicated code around genesis base fees for EIP-1559). Also fixes some bugs with Paris activation checks

Incomplete right now:

- [x] Update calls to `ChainSpec` at callsites
- [x] Update docs
- [x] Add validation call in total difficulty stage
- [x] Add more test vectors for validation functions, fork IDs and fork filters

### CODE REVIEW

**Comment by gakonst** [comment]
a few conflicts on this. also if the chainspecs are not needed either way we can pull them to a separate pr

File: crates/consensus/src/validation.rs
Comment (gakonst) [inline crates/consensus/src/validation.rs:247]: is this relevant? what check needs to exist here, given that we now check total iff inside of the HeadersTD stage? cc @rkrasiuk

File: crates/storage/provider/src/utils.rs
Comment (gakonst) [inline crates/storage/provider/src/utils.rs:31]: Just to make sure I understand:
* If it has block reward, it means it's pre-merge, as a result the TotalDifficulty stage will populate that value and we can leave it to 0
* If it has no block reward, it's post merge, so TD doesn't go up anymore, so it's at `58_750_000_000_000_000_000_000u128`

Should we have a parameter for the max block the TD stage should be running until, optionally, for merged networks?

File: crates/stages/src/stages/total_difficulty.rs
Comment (gakonst) [inline crates/stages/src/stages/total_difficulty.rs:76]: Ah I see, this is the check which we said we'd have to pull out to the TD stage. Makes sense. I was hoping we could maybe skip running this stage for blocks after the merge, but makes sense to do this so that we run the validation rules. It does mean that we do validation in 2 different places (one in consensus, one here).

File: crates/rpc/rpc-engine-api/src/engine_api.rs
Comment (gakonst) [inline crates/rpc/rpc-engine-api/src/engine_api.rs:158]: think this may work too

```suggestion
        let Some(parent_td) = self.client.header_td(&block.parent_hash)? else {
```

**Review by mattsse — APPROVED** [review]
do we no longer need the json files?

File: crates/stages/src/stages/bodies.rs
Comment (gakonst) [inline crates/stages/src/stages/bodies.rs:162]: Meh so on both Bodies + Execution we need to make 1 extra query to the TD stage for each block to make sure that we apply the correct block reward based on the TD.

File: crates/primitives/src/forkid.rs
Comment (gakonst) [inline crates/primitives/src/forkid.rs:354]: i see so this may trigger it to actually be a `ForkFilter::Time` if the active key is that

File: crates/primitives/src/forkid.rs
Comment (gakonst) [inline crates/primitives/src/forkid.rs:271]: The type information is lost from the EIP eh? so we really need to make an assumption that when `head_block_or_time` is `ForkFilterKey::Block` that `fork_id.next` is gonna correspond to a block value, and correspondingly for `Time`?

File: crates/primitives/src/forkid.rs
Comment (gakonst) [inline crates/primitives/src/forkid.rs:75]: where? seems fine

File: crates/interfaces/src/executor.rs
Comment (gakonst) [inline crates/interfaces/src/executor.rs:19]: I really don't like how we're polluting the interface with both the `senders` and the `total_difficulty` option.. I think fine for now but e.g. if I was consuming this client-side, I'd make it so that `BlockExecutor` takes a block where transactions + senders are in 1 struct, and the total_difficulty param is not needed as the executor has some stat var over the block number to check instead of TD

**Review by gakonst — APPROVED** [review]
Took a close look - LGTM and nice work on the test vectors. Feel free to merge unless anything else you'd like to handle.

**Review by gakonst — COMMENTED** [review]
The genesis jsons under bin/reth were dupes of https://github.com/leruaa/reth/tree/6f52f3912b2aa504c0ddef7e2f62a7e285eda1dc/crates/primitives/res/genesis

File: crates/interfaces/src/executor.rs
Comment (onbjerg) [reply crates/interfaces/src/executor.rs:19]: We used to use a type for tx with recovered senders but it was removed from all interfaces in a PR for some reason

**Pull request merged**
